os/board/rtl8721csm/src/component/os/tizenrt/rtk_netmgr.c

```diff
@@ -83,15 +83,81 @@ static wifi_utils_scan_list_s *g_scan_list;
 static int g_scan_num;
 extern struct netdev *ameba_nm_dev_wlan0;
 
+struct ap_scan_list {
+	unsigned int channel;
+	char ssid[WIFI_UTILS_SSID_LEN + 1];
+	unsigned int ssid_length;
+	int rssi;
+	wifi_utils_ap_auth_type_e ap_auth_type;
+	wifi_utils_ap_crypto_type_e ap_crypto_type;
+};
+typedef struct ap_scan_list ap_scan_list_s;
+
+static ap_scan_list_s *saved_scan_list;
+static uint32_t scan_number = 0;
+uint32_t ap_channel;
+
+#include "freertos/wrapper.h"
+#define SCAN_TIMER_DURATION 180000
+static _timer scan_timer;
+
+static void _scan_timer_handler(void *FunctionContext)
+{
+	ndbg("scan Timer expired : release saved scan list\r\n");
+	ndbg("scan Timer expired : sizeof(ap_scan_list_s) =%d, scan_number=%d \r\n", sizeof(ap_scan_list_s), scan_number);
+
+	rtw_mfree((unsigned char *)saved_scan_list, sizeof(ap_scan_list_s) * scan_number);
+	scan_number = 0;
+	rtw_del_timer(&(scan_timer));
+}
+
+static int save_scan_list(wifi_utils_scan_list_s *p_scan_list)
+{
+
+	// If application calls scan before scan result free(before timeout), release scan list and cancel timer
+	if (scan_number) {
+		rtw_mfree((unsigned char *)saved_scan_list, sizeof(ap_scan_list_s) * scan_number);
+		rtw_cancel_timer(&(scan_timer));
+		ndbg("scan is called before timeout\r\n");
+	} else {
+		rtw_init_timer(&(scan_timer), &(scan_timer), _scan_timer_handler, &(scan_timer), "dynamic_chk_timer");
+		ndbg("Start scan timer\n");
+	}
+
+	rtw_set_timer(&(scan_timer), SCAN_TIMER_DURATION);
+

```

Comment:
>  if driver callback(handle scan result) is executing save_scanlist and it writes to saved_scanlist, while timer run scan_timer_handle and free saved_scanlist. then it cause unexpected behavior. So you should protect saved_scanlist. Please check the case.

@pillip8282 , The `saved_scan_list` struct has been used in three cases 
1) The write is done inside `save_scan_list()`, but we start the timer with `SCAN_TIMER_DURATION` before every write.
So there won't be immediate any timeout which will free the struct to interrupt the write function and in my opinion there is no need to protect the write to `saved_scan_list` struct. what is your suggestion?

2) The search is done inside `wifi_netmgr_utils_connect_ap()` and there is chance of timeout handler to free the struct during search. We already planned to add mutex to protect the `saved_scan_list` here but we would to like to get suggestion on adding `rtw_mutext_init` where it needs to be called only once before using `rtw_mutext_get()` and `rtw_mutext_put()`. How about mutex init during `wifi_on()`?

3) The free is done inside `_scan_timer_handler()`.

Reply:
* About case 1 consider this scenario.
1) wifi_manager requests scan.
2) a driver calls save_scan_list and set a timer
3) after 3minutes later, a timer is expired. so the timer execute rtw_mfree in _scan_timer_handler().
4) context switching, wifi_manager request scan.
5) driver callback execute save_scan_list, and _scan_timer_handler wasn't finished it'll call rtw_mfree again.

* I think that initializing a mutex in wifi_netmgr_utils_init(struct netdev *dev) and deinit it in wifi_netmgr_utils_deinit()

Reply:
@pillip8282, Ok I understand the scenario and will implement mutex as per your suggestion.

Reply:
Done.